contracts/TopShotMarketV3.cdc

```diff
@@ -0,0 +1,341 @@
+                    let token <- v1MarketRef.withdraw(tokenID: tokenID)
+
+                    // borrow a reference to the main top shot collection
+                    let ownerCollectionRef = self.ownerCollection.borrow()
+                        ?? panic("Could not borrow owner collection reference")
+
+                    // deposit the withdrawn moment into the main collection
+                    ownerCollectionRef.deposit(token: <-token)
+                } else {
+                    panic("Token with the specified ID is not already for sale in either sale collection")
+                }
+
+            } else {
+                panic("Token with the specified ID is not already for sale in either sale collection")
+            }
+        }
+
+        // purchase lets a user send tokens to purchase an NFT that is for sale
+        // the purchased NFT is returned to the transaction context that called it
+        //
+        // Parameters: tokenID: the ID of the NFT to purchase
+        //             buyTokens: the fungible tokens that are used to buy the NFT
+        //
+        // Returns: @TopShot.NFT: the purchased NFT
+        pub fun purchase(tokenID: UInt64, buyTokens: @FungibleToken.Vault): @TopShot.NFT {
+
+            if self.prices[tokenID] != nil {
+                assert(
+                    buyTokens.balance == self.prices[tokenID]!,
+                    message: "Not enough tokens to buy the NFT!"
+                )
+
+                // Read the price for the token
+                let price = self.prices[tokenID]!
+
+                // Set the price for the token to nil
+                self.prices[tokenID] = nil
+
+                // Take the cut of the tokens that the beneficiary gets from the sent tokens
+                let beneficiaryCut <- buyTokens.withdraw(amount: price*self.cutPercentage)
+
+                // Deposit it into the beneficiary's Vault
+                self.beneficiaryCapability.borrow()!
+                    .deposit(from: <-beneficiaryCut)
+                
+                // Deposit the remaining tokens into the owners vault
+                self.ownerCapability.borrow()!
+                    .deposit(from: <-buyTokens)
+
+                emit MomentPurchased(id: tokenID, price: price, seller: self.owner?.address)
+
+                // Return the purchased token
+                let boughtMoment <- self.ownerCollection.borrow()!.withdraw(withdrawID: tokenID) as! @TopShot.NFT
+
+                return <-boughtMoment
+
+            } else {
+
+                if let v1Market = self.marketV1Capability {
+                    let v1MarketRef = v1Market.borrow()!
+
+                    return <-v1MarketRef.purchase(tokenID: tokenID, buyTokens: <-buyTokens)
+                } else {
+                    panic("No token matching this ID for sale!")
+                }
+            }
+            
+            destroy buyTokens // THIS LINE NEEDS TO BE REMOVED ONCE THE BUG IN CADENCE IS FIXED
```

Comment:
Just for context - what is the bug in Cadence?

Reply:
https://github.com/onflow/cadence/pull/1000
It'll be released in the next version of Cadence, and I'll update this after that is released